src/ICSharpCode.SharpZipLib/Zip/ZipInputStream.cs

```diff
@@ -522,27 +576,67 @@ private int InitialRead(byte[] destination, int offset, int count)
 					throw new ZipException("No password set.");
 				}
 
-				// Generate and set crypto transform...
-				var managed = new PkzipClassicManaged();
-				byte[] key = PkzipClassic.GenerateKeys(ZipStrings.ConvertToArray(password));
+				if (entry.AESKeySize == 0)
+				{
+					// Generate and set crypto transform...
+					var managed = new PkzipClassicManaged();
+					byte[] key = PkzipClassic.GenerateKeys(ZipStrings.ConvertToArray(password));
 
-				inputBuffer.CryptoTransform = managed.CreateDecryptor(key, null);
+					inputBuffer.CryptoTransform = managed.CreateDecryptor(key, null);
 
-				byte[] cryptbuffer = new byte[ZipConstants.CryptoHeaderSize];
-				inputBuffer.ReadClearTextBuffer(cryptbuffer, 0, ZipConstants.CryptoHeaderSize);
+					byte[] cryptbuffer = new byte[ZipConstants.CryptoHeaderSize];
+					inputBuffer.ReadClearTextBuffer(cryptbuffer, 0, ZipConstants.CryptoHeaderSize);
 
-				if (cryptbuffer[ZipConstants.CryptoHeaderSize - 1] != entry.CryptoCheckValue)
-				{
-					throw new ZipException("Invalid password");
-				}
+					if (cryptbuffer[ZipConstants.CryptoHeaderSize - 1] != entry.CryptoCheckValue)
+					{
+						throw new ZipException("Invalid password");
+					}
 
-				if (csize >= ZipConstants.CryptoHeaderSize)
-				{
-					csize -= ZipConstants.CryptoHeaderSize;
+					if (csize >= ZipConstants.CryptoHeaderSize)
+					{
+						csize -= ZipConstants.CryptoHeaderSize;
+					}
+					else if (!usesDescriptor)
+					{
+						throw new ZipException($"Entry compressed size {csize} too small for encryption");
+					}
 				}
-				else if (!usesDescriptor)
+				else
 				{
-					throw new ZipException($"Entry compressed size {csize} too small for encryption");
+					int saltLen = entry.AESSaltLen;
+					byte[] saltBytes = new byte[saltLen];
+					int saltIn = inputBuffer.ReadRawBuffer(saltBytes, 0, saltLen);
+
+					if (saltIn != saltLen)
+					{
+						throw new ZipException("AES Salt expected " + saltLen + " got " + saltIn);
+					}
+					
+					//
+					byte[] pwdVerifyRead = new byte[ZipConstants.AESPasswordVerifyLength];
+					int pwdBytesRead = inputBuffer.ReadRawBuffer(pwdVerifyRead, 0, pwdVerifyRead.Length);
+
+					if (pwdBytesRead != pwdVerifyRead.Length)
+					{
+						throw new EndOfStreamException();
+					}
+
+					int blockSize = entry.AESKeySize / 8;   // bits to bytes
+
+					var decryptor = new ZipAESTransform(password, saltBytes, blockSize, false);
+					decryptor.ManualHmac = csize < 0;
+					byte[] pwdVerifyCalc = decryptor.PwdVerifier;
+					if (pwdVerifyCalc[0] != pwdVerifyRead[0] || pwdVerifyCalc[1] != pwdVerifyRead[1])
+					{
+						throw new ZipException("Invalid password for AES");
+					}
+
+					// The AES data has saltLen+AESPasswordVerifyLength bytes as a header, and AESAuthCodeLength bytes
+					// as a footer.
+					csize -= (saltLen + ZipConstants.AESPasswordVerifyLength + ZipConstants.AESAuthCodeLength);
+					inputBuffer.DecryptionLimit = (int)csize;
```

Comment:
Do we want to set DecryptionLimit in cases where we don't know the compressed size? (looks like CalculateDecryptionSize might keep subtracting values from it, even if it's negative to begin with?)

Reply:
If it is set to a negative value it is behaving the same as when it is set to null.

But I set it to null now and removed the check for a negative value in the InflaterInputStream instead.